### DISCUSSION TITLE
Airflow plugin doesn't respect SQLALCHEMY_BINDS

### DISCUSSION THREAD

**Discussion by wills-feng** [apache/airflow#20080] [Q&A]
My plugin needs to connect to an external database, so I defined SQLALCHEMY_BINDS in webserver_config.py and set __bind_key__ in my model, but it still connects to the core database.

I confirm the SQLALCHEMY_BINDS has been set up in app.config. Does anyone have a solution for this? Thanks.

**Comment by mlem8** [comment]
+1 
I am experiencing the same behavior with my custom airflow plugin

**Comment by JoeIlluminati** [comment]
I have also been experiencing this behavior in airflow 2.3.0.  I did some digging around and was able to pinpoint where it swaps in the default airflow db connection to [here](https://github.com/apache/airflow/blob/2.3.0/airflow/www/extensions/init_appbuilder.py#L331-L336) (confirmed that this snippet is unchanged in the latest version as well).
```python
    def _check_and_init(self, baseview):
        if hasattr(baseview, 'datamodel'):
            baseview.datamodel.session = self.session
        if hasattr(baseview, "__call__"):
            baseview = baseview()
        return baseview
```

`self.session` in this case is the default airflow db conn.  I was able to work around this by defining an `__init__` on my `ModelView` subclass where I would reset the `datamodel` session to my external db conn AND passing the view to my custom plugin as the class itself, not an instantiated object.  Here's an example:

```python
from airflow.plugins_manager import AirflowPlugin
from airflow.www.utils import CustomSQLAInterface
from flask_appbuilder import ModelView
from sqlalchemy import create_engine
from sqlalchemy.orm import scoped_session, sessionmaker

from model import MyCustomModel

external_db_session = scoped_session(sessionmaker(bind=create_engine("<db conn>")))

class MyCustomModelView(ModelView):
    # The second arg is what gets swapped in the _check_and_init call
    datamodel = CustomSQLAInterface(MyCustomModel, external_db_session)

    ...

    def __init__(self, **kwargs):
        # Hacky way to reset the session here
        self.datamodel = CustomSQLAInterface(MyCustomModel, external_db_session)

        super().__init__(**kwargs)

# Note that we are passing in the class and not an object of type of that class.
# This is necessary to ensure that _check_and_init will instantiate it, letting us
# reset the datamodel to have the session we want it to have.
my_custom_view = {
    "name": "Custom View",
    "category": "Custom",
    "view": MyCustomModelView
}

class CustomPlugin(AirflowPlugin):
    name = "custom_plugin"
    appbuilder_views = [my_custom_view]
```

I am unsure if forcing the use of airflows db session is default behavior or not.  I could not find any documentation or discussion indicating it one way or the other.  If a maintainer could provide some insight here that would be awesome.  If this in in fact a bug I'm happy to take a stab at putting up a proper fix.

**Reply by potiuk** [reply]
I think it has never been discussed. It's purely accidental. If you come up with changing it, it would be great. I think the case of having multiple SQLAlchemy  connections for Airlfow is an edge case that very, very few people use, so it's neither discussed nor documented. 

But if you can test, provide PR including documenting the use of it - you are most welcome.

**Reply by JoeIlluminati** [reply]
Thanks for the reply @potiuk.  I can work on taking a crack at that.
